xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,377 @@
+            SymbolTable.lookup_symbol(self, self.nextState), State
+        ):
+            raise VerifyException("Can not find next state")
+        if self.guard.blocks and not isinstance(self.guard.block.last_op, Return):
+            raise VerifyException("Guard region must terminate with ReturnOp")
+        var = self.parent_op()
+        assert isinstance(var, State)
+        if var.transitions != self.parent_region():
+            raise VerifyException("Transition must be located in a transitions region")
+
+
+@irdl_op_definition
+class Machine(IRDLOperation):
+    """Represents a finite-state machine, including a machine name,
+    the type of machine state, and the types of inputs and outputs. This op also
+    includes a `$body` region that contains internal variables and states."""
+
+    name = "fsm.machine"
+
+    body: Region = region_def()
+
+    sym_name = attr_def(StringAttr)
+    initialState = attr_def(StringAttr)
+    function_type = attr_def(FunctionType)
+    arg_attrs = opt_attr_def(ArrayAttr[DictionaryAttr])
+    res_attrs = opt_attr_def(ArrayAttr[DictionaryAttr])
+    arg_names = opt_attr_def(ArrayAttr[StringAttr])
+    res_names = opt_attr_def(ArrayAttr[StringAttr])
+
+    traits = frozenset([NoTerminator(), SymbolTable()])
+
+    def __init__(
+        self,
+        sym_name: str,
+        initial_state: str,
+        function_type: FunctionType | tuple[Sequence[Attribute], Sequence[Attribute]],
+        arg_attrs: ArrayAttr[DictionaryAttr] | None,
+        res_attrs: ArrayAttr[DictionaryAttr] | None,
+        arg_names: ArrayAttr[StringAttr] | None,
+        res_names: ArrayAttr[StringAttr] | None,
+        body: Region | type[Region.DEFAULT] = Region.DEFAULT,
+    ):
+        attributes: dict[str, Attribute | None] = {}
+        attributes["sym_name"] = StringAttr(sym_name)
+        attributes["initialState"] = StringAttr(initial_state)
+        if isinstance(function_type, tuple):
+            inputs, outputs = function_type
+            function_type = FunctionType.from_lists(inputs, outputs)
+        attributes["function_type"] = function_type
+        attributes["arg_attrs"] = arg_attrs
+        attributes["res_attrs"] = res_attrs
+        attributes["arg_names"] = arg_names
+        attributes["res_names"] = res_names
+        if not isinstance(body, Region):
+            body = Region(Block())
+
+        super().__init__(attributes=attributes, regions=[body])
+
+    def verify_(self):
+        if not SymbolTable.lookup_symbol(self, self.initialState):
+            raise VerifyException("Can not find initial state")
+        if (
+            self.arg_attrs is not None
+            and self.arg_names is None
+            or self.res_attrs is not None
+            and self.res_names is None
+            or self.arg_attrs is None
+            and self.arg_names is not None
+            or self.res_attrs is None
+            and self.res_names is not None
+        ):
```

Comment:
It feels like this could be implemented in a way that is more understandable. I am thinking of some sort of list of tuples and then check that for all tuples, either both are None or neither is None. At the very least, please add a comment saying that "if {arg, res}_names is given, {arg, res}_attrs is needed as well" or whatever.